[아켄] 1주차 학습 계획 작성

aken-you opened the pull request:
## 구조
![new-stand-structure](https://user-images.githubusercontent.com/96980857/230043837-46156db8-a817-4440-bf85-b000e9eb3525.png)

## 시도했던 부분

1. sass 사용
sass를 이용해 css를 유지보수하기 용이해졌다. `mixin`, `includes`, `&` 등 sass 명령어를 통해 중복되는 부분을 줄일 수 있었다. 하나의 style.css파일로 style.scss를 mapping한 이유는 이미 sass 폴더에 모듈화가 되어있는데 굳이 css에서 각 파일을 mapping할 필요성을 느끼지 못했다. 그리고 하나의 css파일만 존재하면 http 요청 수를 줄일 수 있어서 더 좋지 않을까?
2. postCSS Sorting 확장 프로그램 사용
css의 속성 작성 순서를 지정할 수 있는 프로그램을 사용했다. prettier처럼 정렬되니 관리하기 편해졌다.
3. component 단위로 구현
mvc 패턴은 사용자의 action을 controller에게 전달해 model의 data를 변경하고 변경한 data를 view에게 알려야 하는 구조이다. 큰 어플리케이션이라면 controller의 역할이 많아져 유지보수하기 어려운 코드가 될 것이다. 사용자의 action에 따라 view가 달라져야 하는데 controller를 거치는 게 최선의 방법일까?라는 고민을 하게 됐다.
사용자의 action에 의해 state가 변하고 rerender가 일어나는 구조를 짜고 싶었다. react에서 web component가 이와 같은 동작 원리임을 알았고, js로 web component를 만들어봤다. 모든 component에 공통되는 method가 있지만 전체적인 component틀이 잡히면 Component 클래스를 만들어 각 component가 클래스를 상속받도록 짤 예정이다.
   
## 고민했던 부분

1. sass파일에서 depth가 깊어짐
depth가 깊어지면 그 만큼 가독성이 좋지 않았다. 스타일 지정해야 할 속성이 많다면 해당 class만 따로 작성할 예정이다.
2. 각 component에 있는 `initRender` method가 정말 필요한가?
사실 생성자 함수에 `insertAdjacentElement`(혹은 `insertAdjaceentHTML`)을 실행시켜주고 초기 렌더링이 필요한 부분에 `컴포넌트.render()`를 써주면 되는데, 초기 렌더링과 렌더링되는 동작을 구분하고 싶어 함수명으로 표현했다. 

## 개선해야 할 부분
1. class는 필요할 때만 사용하자
저번 미션에서 굳이 스타일 지정하지 않아도 되는데 class 값을 지정하여 불필요한 class가 있었다. 이번에는 이에 유의하면서 작성할 것이다.
2. component의 기준을 정하자
파일 이름만 봐선 어떤 기준으로 component로 나눈건지 짐작하지 못할 것 같다. 우선 큰 부분(Header, Nav, Main Content)으로 나눠 세부적인 부분은 해당 폴더에 component class 파일을 작성했다. 구조 설계할 땐 component가 되야할 것 같았는데 막상 짜보니 굳이?라는 생각이 들었다. **나만의 component 기준을 세워보자**
3. commit message를 자세하게 남기자
"feat: ~ 클래스 생성"으로 남기니 이 클래스의 역할을 확인하기 위해 직접 파일을 봐야했었다.
**커밋 메세지 본문 내용**: 어떤 역할을 하는 지, 어떤 기능이 추가/삭제/변경 됐는지, 상태

aken-you committed:
chore: eslint, prettier, sass 설치

aken-you committed:
docs: 기획서 기능 목록 정리

aken-you committed:
docs: 구조 사진 추가

realsnoopso commented on readme.md:
```diff
@@ -1 +1,111 @@
+## 구조
 
+![image](https://user-images.githubusercontent.com/96980857/229732784-543b8d21-5593-4efe-ba44-aef3163bf5f4.png)
+
```
![229732784-543b8d21-5593-4efe-ba44-aef3163bf5f4](https://user-images.githubusercontent.com/96381221/229998520-4ac2535e-2f60-4042-b05e-2132c482f985.jpg)

컴포넌트 구조를 짜려는건 아닌것 같긴 한데 일단 생각했을 때 저런 부분들 컴포넌트로 짜서 재 사용하면 좋을 것 같아요

aken-you committed:
feat: sass 폴더 구조 및 폰트&색깔 초기 설정

aken-you committed:
docs: 구조 피드백 반영

aken-you committed:
chore: prettier 규칙 추가 및 package.json에 entry point 설정

aken-you committed:
feat: dom 가져오는 util 함수 생성

aken-you committed:
feat: header 클래스 생성

aken-you committed:
feat: App 클래스 생성

aken-you committed:
feat: entry point인 index.js파일 생성 및 html 파일 생성

aken-you committed:
feat: header 관련 style 적용

aken-you committed:
docs: header 레이아웃 기능 목록 체크

aken-you committed:
feat: MainContentHeader 클래스 생성

- main-content의 header 부분과 관련된 클래스

aken-you committed:
feat: MainContent 클래스 생성

aken-you committed:
feat: main-content header 부분 스타일 적용

aken-you committed:
refactor: scss 파일명 변경

aken-you committed:
feat: App 클래스에 MainContent 클래스 적용

aken-you committed:
refactor: header 링크 설정 및 scss 수정

aken-you committed:
refactor: mainContentHeader의 state 추가

- state: pressTabs와 관련된 정보들 저장

- state를 이용해 press-tab 클래스를 가진 요소의 자식 요소들의 template 생성

JaydenLee1116 commented on .eslintrc.json:
```diff
@@ -0,0 +1,19 @@
+    "no-alert": "off",
+    "prettier/prettier": ["error", { "endOfLine": "auto" }]
+  }
+}
```
아켄의 린트 설정ㅋㅋㅋ 저는 복잡한 설정은 잘 안건드리는데 이렇게 알고 사용할 수 있으면 정말 좋은 것 같아요!

JaydenLee1116 commented on index.html:
```diff
@@ -0,0 +1,14 @@
+  </head>
+  <body>
+    <div id="app"></div>
+    <script type="module" src="./src/js/index.js"></script>
```
script 태그를 head에 안두시고 body 맨 아래에 두신 이유가 따로 있을까요?!

aken-you replied:
head 태그 안에서 defer를 이용해 script 태그를 놓는 방법도 있지만 [일부 브라우저에서 defer 속성을 지원해주지 않아서](https://caniuse.com/?search=defer) 우선 body태그 최하단에 넣기로 했습니다!

JaydenLee1116 commented on readme.md:
```diff
@@ -1 +1,111 @@
+
+- [ ] 구독 중인 언론사의 브랜드마크 셀에 마우스를 올렸을 때, [구독해지]버튼이 나타난다.
+- [ ] [구독해지]버튼을 클릭했을 때, alert 모달창이 나타난다.
+- [ ] 구독을 해지했을 경우, 해당 언론사의 브랜드 마크가 사라진다.
```
피쳐 리스트 진짜 자세하게 잘 쓰신 거 같아요. 저도 본받아서 좀더 다듬어 봐야겠어요!!!

JaydenLee1116 commented on src/css/style.css.map:
```diff
@@ -0,0 +1,5 @@
+{"version":3,"sourceRoot":"","sources":["../sass/base/_typography.scss","../sass/base/_reset.scss","../sass/abstracts/_mixins.scss","../sass/abstracts/_variables.scss","../sass/layout/_app.scss","../sass/layout/_header.scss","../sass/layout/_main-content.scss"],"names":[],"mappings":"AAAQ;ACAR;EACE;EACA;EACA;;;AAGF;EACE;EACA;;;AAGF;EACE;;;ADVF;EACE;;;AAIA;EENA;EACA;EACA;;AFOA;EETA;EACA;EACA;EFSE, ...
+=======
+{"version":3,"sourceRoot":"","sources":["../sass/base/_typography.scss","../sass/base/_reset.scss","../sass/abstracts/_variables.scss","../sass/abstracts/_mixins.scss","../sass/layout/_app.scss","../sass/layout/_header.scss","../sass/layout/_main-content.scss"],"names":[],"mappings":"AAAQ;ACAR;EACE;EACA;EACA;;;AAGF;EACE;;;AAGF;EACE;;;ADTF;EACE;EACA,OEIM;;;AFAN;EGPA;EACA;EACA;;AHQA;EGVA;EACA;EACA; ...
+>>>>>>> Stashed changes
```
이렇게 컴파일 시에 생기는 파일은 gitignore에서 처리해서 commit 되지 않게 해줘도 좋을 거 같아요!

JaydenLee1116 commented on src/js/components/header.js:
```diff
@@ -0,0 +1,34 @@
+  const weekday = ['일', '월', '화', '수', '목', '금', '토'][data.getDay()]; // 0부터 시작하므로 1을 더하고 문자열 형태로 변환
+  return `${year}.${month}.${day} ${weekday}요일`;
+};
+
```
이 함수는 따로 빼서 유틸함수처럼(?) 사용할 수 있게 하면 더 좋을 거 같아요!

aken-you replied:
지금까지 Header 클래스에서만 사용해서 여기에 위치해뒀습니다! 나중에 다른 곳에서도 쓰이면 그때 utils 폴더에 넣으려구요!

JaydenLee1116 commented on src/js/components/header.js:
```diff
@@ -0,0 +1,34 @@
+
+  initRender() {
+    this.$parent.insertAdjacentHTML('afterbegin', this.template());
+  }
```
```js
initRender() {
    this.$parent.insertAdjacentHTML('afterbegin', this.template());
  }

initRender(position = 'afterbegin') {
    this.$parent.insertAdjacentHTML(position, this.template());
  }
```
위와 같이 위치를 파라미터로 전달하되, 'afterbegin'을 기본값으로 해서 좀더 붙이는 위치를 유연하게 가져가도 좋지 않을까 조심스레 제안해봅니당!

aken-you replied:
아무래도 저 position이 하드코딩이라서 수정하려고 했는데 의견 감사합니다! 공통된 Component 클래스 만들 때 파라미터 기본값을 설정해줘야겠네요 😄

JaydenLee1116 commented on src/js/components/header.js:
```diff
@@ -0,0 +1,34 @@
+        </a>
+        <div class="header__today">${getDateContent(today)}</div>
+      </header>
+    `;
```
어떤 탬플릿을 반환하니까, getter를 써서 표현해도 좋을 것 같아요! 혹은 `getTemplate()`은 어떨까요!

aken-you replied:
태그 안에 있는 `textContent`를 반환하는 함수라서 `getDateContent`라고 지었습니다!

JaydenLee1116 commented on src/js/components/mainContent/mainContent.js:
```diff
@@ -0,0 +1,16 @@
+  constructor($parent) {
+    this.$parent = $parent;
+    this.$ele = document.createElement('main');
+    this.$ele.id = 'main-content';
```
여기도 Header class처럼 template을 만드는 함수를 만들어서 통일성있게 해줘도 좋을 것 같아요!

aken-you replied:
다시 보니 통일성을 위해 template 함수를 미리 써놓고 이후에 수정하면 좋았을 것 같네요!

JaydenLee1116 commented on src/js/utils/dom.js:
```diff
@@ -0,0 +1,4 @@
+export const $ = ({ selector, parent = document }) =>
+  parent.querySelector(selector);
+export const $$ = ({ selector, parent = document }) =>
+  parent.querySelectorAll(selector);
```
옹... 저는 늘 기본으로 document에서 찾게 했는데, 이렇게 하면 조금더 짧은 경로로 원하는 노드를 찾을 수 있겠네요!

JaydenLee1116 commented on src/js/components/header.js:
```diff
@@ -0,0 +1,34 @@
+    return String.raw`
+      <header id="header">
+        <a href="/" class="header__logo">
+          <img src="../../src/images/logo.svg">
```
뭔가 여기 src값으로 들어가는 경로를 따로 빼주면 좋을 것 같은데, 저도 어떤 방법이 좋을지 생각해보겠습니다!

aken-you replied:
다시 보니 경로가 수정되었을 때 고치기 힘들 수 있을 것 같네요! private field로 관리하는 방법도 있을 것 같아요 :)

JaydenLee1116 reviewed (commented):
아켄켄~ 확실히 페어리뷰하면서 배우는 게 많은 거 같아요ㅋㅋㅋㅋㅋ
담주도 화이팅입니다!!!

crongro commented:
> * sass 사용
>   sass를 이용해 css를 유지보수하기 용이해졌다. `mixin`, `includes`, `&` 등 sass 명령어를 통해 중복되는 부분을 줄일 수 있었다. 하나의 style.css파일로 style.scss를 mapping한 이유는 이미 sass 폴더에 모듈화가 되어있는데 굳이 css에서 각 파일을 mapping할 필요성을 느끼지 못했다. 그리고 하나의 css파일만 존재하면 http 요청 수를 줄일 수 있어서 더 좋지 않을까?
> * postCSS Sorting 확장 프로그램 사용
>   css의 속성 작성 순서를 지정할 수 있는 프로그램을 사용했다. prettier처럼 정렬되니 관리하기 편해졌다.
> * component 단위로 구현
>   mvc 패턴은 사용자의 action을 controller에게 전달해 model의 data를 변경하고 변경한 data를 view에게 알려야 하는 구조이다. 큰 어플리케이션이라면 controller의 역할이 많아져 유지보수하기 어려운 코드가 될 것이다. 사용자의 action에 따라 view가 달라져야 하는데 controller를 거치는 게 최선의 방법일까?라는 고민을 하게 됐다.
>   사용자의 action에 의해 state가 변하고 rerender가 일어나는 구조를 짜고 싶었다. react에서 web component가 이와 같은 동작 원리임을 알았고, js로 web component를 만들어봤다. 모든 component에 공통되는 method가 있지만 전체적인 component틀이 잡히면 Component 클래스를 만들어 각 component가 클래스를 상속받도록 짤 예정이다.

고민을 하는것도, 방향도 아주 좋군요.

crongro commented on src/js/app.js:
```diff
@@ -0,0 +1,18 @@
+import Header from './components/header.js';
+import MainContent from './components/mainContent/mainContent.js';
+
+export default class App {
+  constructor($targetEle) {
+    this.$targetEle = $targetEle;
+
+    this.header = new Header(this.$targetEle);
+    this.mainContent = new MainContent(this.$targetEle);
+
+    this.initRender();
+  }
+
+  initRender() {
+    this.header.initRender();
+    this.mainContent.initRender();
+  }
+}
```
이 클래스는 그냥 함수로 해도 될 듯.
App을 누군가 다른곳에서 부를일도 없고, (app 객체를 여러개 만들일은 없죠)
메서드도 현재 한개 뿐이기도하고요.

crongro commented on src/js/components/mainContent/mainContentHeader.js:
```diff
@@ -0,0 +1,42 @@
+    this.$ele = document.createElement('header');
+    this.$ele.className = 'main-content__header';
+
+    this.state = {
+      pressTabs: [
+        { contents: '전체 언론사', isClicked: true },
+        { contents: '내가 구독한 언론사', isClicked: false }
+      ]
+    };
```
참고로 상태(state)는 변경이 되는 항목만 상태로 취급해요. 
pressTabs 은 변경되지 않고, 고정된 값이라 state 라기 보다는 그냥 데이터라고 봐야하고요.

crongro commented on src/js/components/mainContent/mainContentHeader.js:
```diff
@@ -0,0 +1,42 @@
+          `<span class="press-tab-btn ${isClicked ? 'active' : ''}">${contents}</span>`
+      )
+      .join('');
+    return String.raw`
```
여기서 string.raw 가 필요한 이유는 뭔가요? 이스케이프 하는 코드가 보이진 않네요.

crongro commented on src/sass/abstracts/_mixins.scss:
```diff
@@ -0,0 +1,12 @@
+  line-height: nth($type, 3);
+}
+
+@mixin flex-row($justify-content: normal, $align-items: normal) {
```
mixin 도 쓰고 좋네요.

crongro reviewed (approved):
구조가 괜찮아보여요.
이상태에서 기능을 계속 늘려보시죠.
